scripts/extensions/sams/src/components/sets/setEditorPanel.tsx

```diff
@@ -0,0 +1,245 @@
+
+// UI
+import {Button, ButtonGroup, Input, Option, Select, Switch, FormLabel} from 'superdesk-ui-framework/react';
+import {
+    FormGroup,
+    FormRow,
+    PanelContent,
+    PanelContentBlock,
+    PanelContentBlockInner,
+    PanelHeader,
+    PanelHeaderSlidingToolbar,
+    Text,
+} from '../../ui';
+
+// Utils
+import {hasItemChanged} from '../../utils/api';
+
+interface IProps {
+    original?: ISetItem;
+    destinations: Array<IStorageDestinationItem>;
+    closeEditor(): void;
+    previewSet(set: ISetItem): void;
+    updateSet(original: ISetItem, updates: Partial<ISetItem>): Promise<ISetItem>;
+    createSet(set: Partial<ISetItem>): Promise<ISetItem>;
+    currentDestination?: IStorageDestinationItem;
+}
+
+interface IState {
+    updates: Partial<ISetItem>;
+    isDirty?: boolean;
+}
+
+export function getSetEditorPanel(superdesk: ISuperdesk) {
+    const {gettext} = superdesk.localization;
+
+    const mapStateToProps = (state: IApplicationState) => ({
+        original: getSelectedSet(state),
+        destinations: getStorageDestinations(state),
+        currentDestination: getSelectedSetStorageDestination(state),
+    });
+
+    const mapDispatchToProps = (dispatch: Dispatch) => ({
+        closeEditor: () => dispatch(closeSetContentPanel()),
+        previewSet: (set: ISetItem) => dispatch(previewSet(set._id)),
+        updateSet: (original: ISetItem, updates: ISetItem) => dispatch<any>(updateSet(original, updates)),
+        createSet: (set: ISetItem) => dispatch<any>(createSet(set)),
+    });
+
+    class SetEditorPanelComponent extends React.Component<IProps, IState> {
+        onChange: Dictionary<string, (value: any) => void>;
+
+        constructor(props: IProps) {
+            super(props);
+
+            if (this.props.original?._id == null) {
+                this.state = {
+                    updates: {
+                        destination_name: this.props.destinations?.[0]?._id,
+                    },
+                    isDirty: true,
+                };
+            } else {
+                this.state = {
+                    updates: cloneDeep<ISetItem>(this.props.original),
+                    isDirty: false,
+                };
+            }
+
+            this.onStateChange = this.onStateChange.bind(this);
+            this.onSave = this.onSave.bind(this);
+            this.onCancel = this.onCancel.bind(this);
+
+            this.onChange = {
+                name: (value: string) => this.onFieldChange('name', value),
+                description: (value: string) => this.onFieldChange('description', value),
+                destination_name: (value: string) => this.onFieldChange('destination_name', value),
+                state: (value: boolean) => this.onStateChange(value),
+            };
+        }
+
+        onFieldChange(field: keyof ISetItem, value: any): void {
+            const updates = this.state.updates;
+            let dirty = true;
```

Comment:
What about computing dirtiness in the render method? It would allow dropping `this.onChange` and `onFieldChange` which don't seem to be doing anything, but passing values around.

Reply:
1) It is better to calculate the `dirty` state only when the attributes change, not on every single render
2) The updates are stored in the local state, and is used when creating or updating an item. The untouched original item is passed in through `props.original` and the updates are stored in `state.updates`
https://github.com/superdesk/superdesk-client-core/blob/8bee669e05cadd9db6c1490ec4a82086d71c328a/scripts/extensions/sams/src/components/sets/setEditorPanel.tsx#L130
https://github.com/superdesk/superdesk-client-core/blob/8bee669e05cadd9db6c1490ec4a82086d71c328a/scripts/extensions/sams/src/components/sets/setEditorPanel.tsx#L132

Reply:
My bad, I thought dirtiness was checked via reference equality. Having `onChange` there still doesn't seem useful compared to calling `onFieldChange`/`onStateChange` directly when handling input changes.

Reply:
By convention I do not use anonymous functions in the render function. Using the `onChange` object provides the same reference to the callback on every render

Reply:
> onChange: Dictionary<string, (value: any) => void>;

Yep, I agree that having the same references is useful for checking reference equality. Could you declare it as separate functions so we don't lose type safety due to `any` type?